sql-plugin/src/main/311until340-all/scala/org/apache/spark/sql/rapids/arithmetic.scala

```diff
@@ -0,0 +1,246 @@
+        xyXor.bitAnd(xrXor)
+      }
+    }
+    val signDiffCV = withResource(signCV) { sign =>
+      withResource(Scalar.fromInt(0)) { zero =>
+        sign.lessThan(zero)
+      }
+    }
+    withResource(signDiffCV) { signDiff =>
+      withResource(signDiff.any()) { any =>
+        if (any.isValid && any.getBoolean) {
+          throw RapidsErrorUtils.arithmeticOverflowError(
+            "One or more rows overflow for Subtract operation."
+          )
+        }
+      }
+    }
+  }
+
+  private[this] def decimalOpOverflowCheck(
+      lhs: BinaryOperable,
+      rhs: BinaryOperable,
+      ret: ColumnVector): ColumnVector = {
+    // We need a special overflow check for decimal because CUDF does not support INT128 so we
+    // cannot reuse the same code for the other types.
+    // Overflow happens if the arguments have different signs and the sign of the result is
+    // different from the sign of subtractend (RHS).
+    val numRows = ret.getRowCount.toInt
+    val zero = BigDecimal(0).bigDecimal
+    val overflow = withResource(DecimalUtils.lessThan(rhs, zero, numRows)) { rhsLz =>
+      val argsSignDifferent = withResource(DecimalUtils.lessThan(lhs, zero, numRows)) { lhsLz =>
+        lhsLz.notEqualTo(rhsLz)
+      }
+      withResource(argsSignDifferent) { argsSignDifferent =>
+        val resultAndSubtrahendSameSign =
+          withResource(DecimalUtils.lessThan(ret, zero)) { resultLz =>
+            rhsLz.equalTo(resultLz)
+          }
+        withResource(resultAndSubtrahendSameSign) { resultAndSubtrahendSameSign =>
+          resultAndSubtrahendSameSign.and(argsSignDifferent)
+        }
+      }
+    }
+    withResource(overflow) { overflow =>
+      if (failOnError) {
+        withResource(overflow.any()) { any =>
+          if (any.isValid && any.getBoolean) {
+            throw new ArithmeticException("One or more rows overflow for Subtract operation.")
+          }
+        }
+        ret.incRefCount()
+      } else {
+        withResource(GpuScalar.from(null, dataType)) { nullVal =>
+          overflow.ifElse(nullVal, ret)
+        }
+      }
+    }
+  }
+
+  override def doColumnar(lhs: BinaryOperable, rhs: BinaryOperable): ColumnVector = {
+    val ret = super.doColumnar(lhs, rhs)
+    withResource(ret) { ret =>
+      // No shims are needed, because it actually supports ANSI mode from Spark v3.0.1.
+      if (failOnError && GpuAnsi.needBasicOpOverflowCheck(dataType) ||
+          GpuTypeShims.isSupportedDayTimeType(dataType) ||
+          GpuTypeShims.isSupportedYearMonthType(dataType)) {
+        // For day time interval, Spark throws an exception when overflow,
+        // regardless of whether `SQLConf.get.ansiEnabled` is true or false
+        basicOpOverflowCheck(lhs, rhs, ret)
+      }
+
+      if (dataType.isInstanceOf[DecimalType]) {
+        decimalOpOverflowCheck(lhs, rhs, ret)
+      } else {
+        ret.incRefCount()
+      }
+    }
+  }
+}
+
+case class GpuMultiply(
```

Comment:
Why do we shim GpuMultiply if it does not handle decimal at all . Did something else change that makes it so we need to have multiple copies of it?

Reply:
Sorry, this was an oversight and misunderstanding of how we are handling decimal multiply and divide